src/lexer.py

```diff
@@ -0,0 +1,1060 @@
+#######################################
+
+from strings_with_arrows import *
+
+import string
+
+#######################################
+# CONSTANTS
+#######################################
+
+DIGITS = '0123456789'
+LETTERS = string.ascii_letters
+LETTERS_DIGITS = LETTERS + DIGITS
+
+#######################################
+# ERRORS
+#######################################
+
+class Error:
+	def __init__(self, pos_start, pos_end, error_name, details):
+		self.pos_start = pos_start
+		self.pos_end = pos_end
+		self.error_name = error_name
+		self.details = details
+
+	def as_string(self):
+		result  = f'{self.error_name}: {self.details}\n'
+		result += f'File {self.pos_start.fn}, line {self.pos_start.ln + 1}'
+		result += '\n\n' + string_with_arrows(self.pos_start.ftxt, self.pos_start, self.pos_end)
+		return result
+
+class IllegalCharError(Error):
+	def __init__(self, pos_start, pos_end, details):
+		super().__init__(pos_start, pos_end, 'Illegal Character', details)
+
+class ExpectedCharError(Error):
+	def __init__(self, pos_start, pos_end, details):
+		super().__init__(pos_start, pos_end, 'Expected Character', details)
+
+class InvalidSyntaxError(Error):
+	def __init__(self, pos_start, pos_end, details=''):
+		super().__init__(pos_start, pos_end, 'Invalid Syntax', details)
+
+class RTError(Error):
+	def __init__(self, pos_start, pos_end, details, context):
+		super().__init__(pos_start, pos_end, 'Runtime Error', details)
+		self.context = context
+
+	def as_string(self):
+		result  = self.generate_traceback()
+		result += f'{self.error_name}: {self.details}'
+		result += '\n\n' + string_with_arrows(self.pos_start.ftxt, self.pos_start, self.pos_end)
+		return result
+
+	def generate_traceback(self):
+		result = ''
+		pos = self.pos_start
+		ctx = self.context
+
+		while ctx:
+			result = f'  File {pos.fn}, line {str(pos.ln + 1)}, in {ctx.display_name}\n' + result
+			pos = ctx.parent_entry_pos
+			ctx = ctx.parent
+
+		return 'Traceback (most recent call last):\n' + result
+
+#######################################
+# POSITION
+#######################################
+
+class Position:
+	def __init__(self, idx, ln, col, fn, ftxt):
+		self.idx = idx
+		self.ln = ln
+		self.col = col
+		self.fn = fn
+		self.ftxt = ftxt
+
+	def advance(self, current_char=None):
+		self.idx += 1
+		self.col += 1
+
+		if current_char == '\n':
+			self.ln += 1
+			self.col = 0
+
+		return self
+
+	def copy(self):
+		return Position(self.idx, self.ln, self.col, self.fn, self.ftxt)
+
+#######################################
+# TOKENS
+#######################################
+
+TT_INT			= 'INT'
+TT_FLOAT    	= 'FLOAT'
+TT_IDENTIFIER	= 'IDENTIFIER'
+TT_KEYWORD		= 'KEYWORD'
+TT_PLUS     	= 'PLUS'
+TT_MINUS    	= 'MINUS'
+TT_MUL      	= 'MUL'
+TT_DIV      	= 'DIV'
+TT_POW			= 'POW'
+TT_EQ			= 'EQ'
+TT_LPAREN   	= 'LPAREN'
+TT_RPAREN   	= 'RPAREN'
+TT_EE			= 'EE'
+TT_NE			= 'NE'
+TT_LT			= 'LT'
+TT_GT			= 'GT'
+TT_LTE			= 'LTE'
+TT_GTE			= 'GTE'
+TT_EOF			= 'EOF'

```

Comment:
Not sure if this is an issue with Github, but, could these be lined up better?

Reply:
For Pycharm users check the following boxes:
* ![Capture](https://user-images.githubusercontent.com/22486123/56532118-350fcf00-6523-11e9-89aa-289eeb2b96f9.PNG)

For Atom users:
[Python-Auto PEP8](https://atom.io/packages/python-autopep8)

Reply:
Its a problem with GitHub, In my code they are aligned.